### Step 1. 필요한 라이브러리 임포트

In [8]:
import pandas as pd
import numpy as np

### Step 2. [데이터](https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv) 확인

- order_id 주문번호
- quantity 주문 개수
- item_name 주문한 상품명
- choice_description 선택사항
- item_price 가격

### Step 3. 데이터 불러오기


**read_csv를 통해 데이터를 올바른 형태로 불러온 뒤 chipo에 저장해주세요**

In [ ]:
!ls
!cd sample_data


sample_data


In [ ]:
cd sample_data

/content/sample_data


In [ ]:
!ls

anscombe.json		      mnist_test.csv
california_housing_test.csv   mnist_train_small.csv
california_housing_train.csv  README.md


In [9]:
from google.colab import files
uploaded = files.upload()

Saving stock_data.csv to stock_data.csv


In [14]:
chipo = pd.read_csv('stock_data.csv', encoding='cp949')

In [115]:
url = 'https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv'
    
# chipo = pd.read_csv('chipotle.tsv', sep = '\t')
chipo = pd.read_csv(url, sep = '\t')

### Step 4. 초반 10행 조회

In [113]:
chipo.head(10)

order_id  ...  item_price
0         1  ...      $2.39 
1         1  ...      $3.39 
2         1  ...      $3.39 
3         1  ...      $2.39 
4         2  ...     $16.98 
5         3  ...     $10.98 
6         3  ...      $1.69 
7         4  ...     $11.75 
8         4  ...      $9.25 
9         5  ...      $9.25 

[10 rows x 5 columns]

In [65]:
len(chipo.groupby('item_name'))

50

In [63]:
chipo[chipo['choice_description'].isna() == False].groupby('choice_description').sum().sort_values('quantity', ascending=False).head(1).index[0]

'[Diet Coke]'

In [43]:
chipo.groupby('item_name').sum().sort_values('quantity', ascending=False).head(1).iloc[0,1]

761

### Step 5. 데이터셋은 몇 개의 행으로 이루어져 있나요? (최소 2가지 이상 풀이법으로 풀어주세요)

4622

In [29]:
alist = list(chipo.columns)
alist

['order_id', 'quantity', 'item_name', 'choice_description', 'item_price']

In [24]:
# Solution 1

chipo.shape[0]  # entries <= 4622 observations

4622

In [25]:
# Solution 2

chipo.info() # entries <= 4622 observations

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            4622 non-null   int64 
 1   quantity            4622 non-null   int64 
 2   item_name           4622 non-null   object
 3   choice_description  3376 non-null   object
 4   item_price          4622 non-null   object
dtypes: int64(2), object(3)
memory usage: 180.7+ KB


### Step 6. 데이터셋은 몇 개의 열로 이루어져 있나요?

5

In [ ]:
chipo.shape[1]

5

### Step 7. 열(column)의 이름들을 모두 조회해주세요

Index(['order_id', 'quantity', 'item_name', 'choice_description',
       'item_price'],
      dtype='object')

In [ ]:
chipo.columns

Index(['order_id', 'quantity', 'item_name', 'choice_description',
       'item_price'],
      dtype='object')

### Step 8. 데이터셋의 인덱스를 조회해주세요

RangeIndex(start=0, stop=4622, step=1)

In [39]:
chipo[chipo.index == 1000]

order_id  ...  item_price
1000       411  ...      $8.49 

[1 rows x 5 columns]

In [ ]:
chipo.index

RangeIndex(start=0, stop=4622, step=1)

### Step 9. 가장 많은 개수가 주문된 item_name은 무엇인가요?

Chicken Bowl	

In [ ]:
c = chipo.groupby('item_name')
c = c.sum()
c = c.sort_values(['quantity'], ascending=False)
c.head(1)

order_id  quantity
item_name                       
Chicken Bowl    713926       761

### Step 10. 가장 많이 주문된 item은 몇 개 주문되었나요?

761

In [ ]:
c = chipo.groupby('item_name')
c = c.sum()
c = c.sort_values(['quantity'], ascending=False)
c.head(1)

order_id  quantity
item_name                       
Chicken Bowl    713926       761

### Step 11. choice_description에서 가장 많이 주문된건 어떤 항목인가요?

Diet Coke 159

In [ ]:
c = chipo.groupby('choice_description').sum()
c = c.sort_values(['quantity'], ascending=False)
c.head(1)
# Diet Coke 159

order_id  quantity
choice_description                    
[Diet Coke]           123455       159

### Step 12. 총 몇 개의 item이 주문되었나요? (quantity의 총 합을 구해주세요)

4972

In [ ]:
total_items_orders = chipo.quantity.sum()
total_items_orders

4972

### Step 13. item price를 float로 변경해주세요

In [66]:
chipo

order_id  ...  item_price
0            1  ...      $2.39 
1            1  ...      $3.39 
2            1  ...      $3.39 
3            1  ...      $2.39 
4            2  ...     $16.98 
...        ...  ...         ...
4617      1833  ...     $11.75 
4618      1833  ...     $11.75 
4619      1834  ...     $11.25 
4620      1834  ...      $8.75 
4621      1834  ...      $8.75 

[4622 rows x 5 columns]

#### Step 13.a. item_price의 데이터타입을 조회해주세요

dtype('O')

In [68]:
chipo.item_price.dtype

dtype('O')

In [ ]:
chipo.item_price.dtype

dtype('O')

#### Step 13.b. item_type에서 $기호를 제거하고 숫자 부분을 float으로 형변환 후 기존 dataframe에 저장해주세요

람다함수를 활용하면 편리합니다

In [92]:
def setfloat(x):
  return float(x[1:-1])

In [116]:
chipo['item_price'] = chipo['item_price'].agg(setfloat)

In [117]:
chipo.head()

order_id  ...  item_price
0         1  ...        2.39
1         1  ...        3.39
2         1  ...        3.39
3         1  ...        2.39
4         2  ...       16.98

[5 rows x 5 columns]

In [ ]:
dollarizer = lambda x: float(x[1:-1])
chipo.item_price = chipo.item_price.apply(dollarizer)

#### Step 13.c. 변경된 item_price의 데이터타입을 조회해주세요

dtype('float64')

In [ ]:
chipo.item_price.dtype

dtype('float64')

### Step 14. 총 매출은 얼마인가요?

39237.02

In [130]:
(chipo['quantity']*chipo['item_price']).sum()

39237.02

In [ ]:
revenue = (chipo['quantity']* chipo['item_price']).sum()

print('Revenue was: $' + str(np.round(revenue,2)))

39237.02

### Step 15. 총 몇 건의 주문이 발생헀나요? (주문번호 기준) 

1834

**4가지 해답을 적어뒀습니다. 이 해답을 참조하여 Step 17을 해결해주세요**

In [137]:
chipo.order_id.nunique()

1834

In [ ]:
# Solution 1
len(chipo['order_id'].unique())

1834

In [ ]:
# Solution 2
chipo['order_id'].unique().shape

(1834,)

In [ ]:
# Solution 3
chipo['order_id'].nunique()

1834

In [139]:
# Solution 4
chipo['order_id'].value_counts().count()

1834

### Step 16. 주문당 평균 매출은 얼마인가요?

21.394231188658654

In [140]:
chipo['revenue'] = chipo['quantity'] * chipo['item_price']
chipo

order_id  quantity  ... item_price revenue
0            1         1  ...       2.39    2.39
1            1         1  ...       3.39    3.39
2            1         1  ...       3.39    3.39
3            1         1  ...       2.39    2.39
4            2         2  ...      16.98   33.96
...        ...       ...  ...        ...     ...
4617      1833         1  ...      11.75   11.75
4618      1833         1  ...      11.75   11.75
4619      1834         1  ...      11.25   11.25
4620      1834         1  ...       8.75    8.75
4621      1834         1  ...       8.75    8.75

[4622 rows x 6 columns]

In [142]:
chipo.groupby('order_id').sum()

quantity  item_price  revenue
order_id                               
1                4       11.56    11.56
2                2       16.98    33.96
3                2       12.67    12.67
4                2       21.00    21.00
5                2       13.70    13.70
...            ...         ...      ...
1830             2       23.00    23.00
1831             3       12.90    12.90
1832             2       13.20    13.20
1833             2       23.50    23.50
1834             3       28.75    28.75

[1834 rows x 3 columns]

In [144]:
order = chipo.groupby(by=['order_id']).sum()
order

quantity  item_price  revenue
order_id                               
1                4       11.56    11.56
2                2       16.98    33.96
3                2       12.67    12.67
4                2       21.00    21.00
5                2       13.70    13.70
...            ...         ...      ...
1830             2       23.00    23.00
1831             3       12.90    12.90
1832             2       13.20    13.20
1833             2       23.50    23.50
1834             3       28.75    28.75

[1834 rows x 3 columns]

In [146]:
order.mean().revenue

21.394231188658654

In [ ]:
# Solution 1

chipo['revenue'] = chipo['quantity'] * chipo['item_price']
order_grouped = chipo.groupby(by=['order_id']).sum()
order_grouped.mean()['revenue']

21.394231188658654

In [ ]:
# Solution 2

chipo.groupby(by=['order_id']).sum().mean()['revenue']

### Step 17. 몇 종류의 item이 판매되었나요?

50

In [ ]:
chipo.item_name.value_counts().count()

50